base/path.jl

```diff
@@ -137,8 +137,9 @@ _splitdir_nodrive(path::String) = _splitdir_nodrive("", path)
 function _splitdir_nodrive(a::String, b::String)
     m = match(path_dir_splitter,b)
     m === nothing && return (a,b)
-    a = string(a, isempty(m.captures[1]) ? m.captures[2][1] : m.captures[1])
-    a, String(m.captures[3])
+    captures = (m.captures...,)::NTuple{3,SubString{String}}
```

Comment:
This probably isn't very performance sensitive, but when I tested this on my machine I got a significant performance regression (630ns -> 770ns). I suspect this line is the cause. Since it would be easy to do this with
```julia
getcapture(v, i) = v[i]::AbstractString
c1, c2, c3 = getcapture(m.captures, 1), getcapture(m.captures, 2), getcapture(m.captures, 3)
```
maybe that would be better?

Reply:
whops, thanks for catching this !

It seems like we can use a closure without problem for this case:
```julia
julia> using BenchmarkTools

julia> @eval Base.Filesystem function _splitdir_nodrive(a::String, b::String)
           m = match(path_dir_splitter,b)
           m === nothing && return (a,b)
           a = string(a, isempty(m.captures[1]) ? m.captures[2][1] : m.captures[1])
           a, String(m.captures[3])
       end
_splitdir_nodrive (generic function with 2 methods)

julia> Base.Filesystem._splitdir_nodrive("", @__DIR__)
("/Users/aviatesk/julia/packages", "JET")

julia> @btime Base.Filesystem._splitdir_nodrive("", @__DIR__);
  829.350 ns (18 allocations: 848 bytes)

julia> @eval Base.Filesystem function _splitdir_nodrive(a::String, b::String)
           m = match(path_dir_splitter,b)
           m === nothing && return (a,b)
           cs = m.captures
           getcapture(cs, i) = cs[i]::SubString{String}
           c1, c2, c3 = getcapture(cs, 1), getcapture(cs, 2), getcapture(cs, 3)
           a = string(a, isempty(c1) ? c2[1] : c1)
           a, String(c3)
       end
_splitdir_nodrive (generic function with 2 methods)

julia> Base.Filesystem._splitdir_nodrive("", @__DIR__)
("/Users/aviatesk/julia/packages", "JET")

julia> @btime Base.Filesystem._splitdir_nodrive("", @__DIR__);
  808.012 ns (18 allocations: 848 bytes)
```

Reply:
That's more specifically an inner function; we could have also called it `getcapture(i) = cs[i]::T` and then it would be a full closure. I didn't benchmark that just because they are sometimes sources of trouble.